server/src/main/java/io/druid/server/security/AuthorizationUtils.java

```diff
@@ -153,25 +151,40 @@ public static Access authorizeAllResourceActions(
   }
 
   /**
-   * Filter a list of resource-actions using the request's authorization fields, returning a new list of
-   * resource-actions that were authorized.
+   * Filter a collection of resources by applying the resourceActionGenerator to each resource, adding
+   * the filtered resources to a caller provided Collection, filteredResources.
+   *
+   * The resourceActionGenerator returns an Iterable<ResourceAction> for each resource.
+   *
+   * If every resource-action in the iterable is authorized, the resource will be added to filteredResources.
+   *
+   * If there is an authorization failure for one of the resource-actions, the resource will not be
+   * added to filteredResources..
+   *
+   * If the resourceActionGenerator returns null for a resource, that resource will not be added to filteredResources.
    *
    * This function will set the DRUID_AUTHORIZATION_CHECKED attribute in the request.
    *
    * If this attribute is already set when this function is called, an exception is thrown.
    *
    * @param request HTTP request to be authorized
-   * @param resources List of resources to be processed into resource-actions
-   * @param resourceActionGenerator Function that creates a resource-action from a resource
-   * @return A list containing the resource-actions from the resourceParser that were successfully authorized.
+   * @param resources resources to be processed into resource-actions
+   * @param resourceActionGenerator Function that creates an iterable of resource-actions from a resource
+   * @param authorizerMapper authorizer mapper
+   * @param filteredResources caller provided collection, resources that pass filtering will be added to this
    */
-  public static <ResType> List<ResType> filterAuthorizedResources(
+  public static <ResType> void filterAuthorizedResources(
       final HttpServletRequest request,
       final Collection<ResType> resources,
-      final Function<? super ResType, ResourceAction> resourceActionGenerator,
-      final AuthorizerMapper authorizerMapper
+      final Function<? super ResType, Iterable<ResourceAction>> resourceActionGenerator,
+      final AuthorizerMapper authorizerMapper,
+      final Collection<? super ResType> filteredResources
```

Comment:
What's the reason for having an out-parameter instead of returning a `Collection<ResType>` or `Iterable<ResType>`? I'm wondering since out-params are harder to comprehend than return values.

I'm thinking that many callers would want to directly use a returned filteredResources, and for the callers that don't, they could do:

```java
Iterables.addAll(filteredResources, filterAuthorizedResources(...))`
```

Reply:
I have the out-parameter so that the caller can control the type of the Collection that gets the filtered values. 

I implemented that way after seeing that SupervisorResource uses Sets for storing supervisor ids, maybe a future caller really cares about using a Set to avoid duplicates, and I thought it'd be better to have the caller pass in the output collection to avoid having to create extra lists and copy them over into other structures.

Reply:
In that case, I think it's best for filterAuthorizedResources to return an `Iterable< ResType>` (and maybe accept one as well, instead of the Collection it accepts now). There won't be any need to create useless collections since the caller can iterate the iterable and directly insert it into the target collection, using `Iterables.addAll`.

Reply:
alright, changed filtereAuthorizedResources to return Iterable